module/zfs/zfs_vnops.c

```diff
@@ -3563,6 +3652,17 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 	}
 
 	/*
+	 * If we are using project inheritance, we only allow renames
+	 * into our tree when the project IDs are the same; otherwise
+	 * tree quota mechanism would be circumvented.
```

Comment:
If the ZFS_PROJINHERIT flag in set on the directory object, then we hope that all the children under the directory will belong to the same project, means the same project ID. Based on that if we want to support tree quota, for example, assume we want to limite the space used by the directory, then we can give an unique project ID to the directory and set ZFS_PROJINHERIT flag on it. And then the project quota mechanism will guarantee that.